client/homebrew/editor/editor.jsx

```diff
@@ -174,9 +178,42 @@ const Editor = createClass({
 		}
 	},
 
-	brewJump : function(){
-		const currentPage = this.getCurrentPage();
-		window.location.hash = `p${currentPage}`;
+	brewJump : function(targetPage=this.getCurrentPage()){
+		if(!window || this.isMeta()) return;
+		// console.log(`Scroll to: p${targetPage}`);
+		window.frames['BrewRenderer'].contentDocument.getElementById(`p${targetPage}`).scrollIntoView({ behavior: 'smooth', block: 'start' });
+		// const hashPage = (page != 1) ? `p${page}` : '';
+		// window.location.hash = hashPage;
+	},
+
+	sourceJump : function(targetLine=null){
+		if(this.isText() || (this.isStyle() && targetLine == 0)) {
+			if(targetLine == null) {
+				targetLine = 0;
+
+				const pageCollection = window.frames['BrewRenderer'].contentDocument.getElementsByClassName('page');
+				const brewRendererHeight = window.frames['BrewRenderer'].contentDocument.getElementsByClassName('brewRenderer').item(0).getBoundingClientRect().height;
+
+				let currentPage = 1;
+				for (const page of pageCollection) {
+					if(page.getBoundingClientRect().bottom > (brewRendererHeight / 2)) {
+						const pageId = page.id;
+						currentPage = parseInt(pageId.slice(1, pageId.length)) || 1;
+						break;
+					}
+				}
+
+				if(currentPage > 1){
+					const textString = this.props.brew.text.split('\\page').slice(0, currentPage-1).join('\\page');
+					const textPosition = textString.length;
+					const lineCount = textString.slice(0, textPosition).split('\n').length;
+
+					targetLine = lineCount;
+				}
```

Comment:
We need to handle the difference between Legacy and V3 `\page` markers again. I might suggest replicating the logic from the `getCurrentPage()` function you made: split the text into lines, and scan through each until the total number of `\page` found matches `currentPage`. Using the same logic also exposes an opportunity for cleanup of duplicate code in the future, where currently it's not obvious that these are doing essentially the same thing.

Reply:
Looks like we can just pass regex to `split()` so it shouldn't be difficult to implement; splitting this to a separate pagination function that can be re-used elsewhere is probably worthwhile for DRY eliminations.

The `currentPage > 1` check was more about it skipping the unnecessary calculations - if current page in the Brew Renderer has been determined to be 1 (or less) then the target line to move to should just remain at it's initial value of 0.

Reply:
> The currentPage > 1 check was more about it skipping the unnecessary calculations...

That's what I figured. But I think in this case the slight improvement in code simplicity is worth the slight chance of being able to skip a calculation.

> splitting this to a separate pagination function that can be re-used elsewhere is probably worthwhile for DRY eliminations.

It would be nice too (in a different PR probably) to have the pagination common between the Editor and BrewRenderer components. As you know already, we do this `currentPage` calculation in `brewRenderer` so we can draw the page count box and handle partial brew rendering.

Reply:
There was a small edge case where the previous logic resulted in the source position for page 1 resulting in line 1, rather than line 0. The current logic checks to see if there are any new lines `\n` in the slice of text between the start and the current page position; if not, then the line is set to 0.
But it occurs to me that this is functionally the same as the `currentPage > 1` check.